File: bindings/php/tests/DataHandleTest.php
<?php

declare(strict_types=1);

/* SPDX-License-Identifier: Apache-2.0 */

namespace Goplasmatic\Datalogic\Tests;

use Goplasmatic\Datalogic\DataHandle;
use Goplasmatic\Datalogic\Engine;
use Goplasmatic\Datalogic\Exception\ParseException;
use PHPUnit\Framework\TestCase;

final class DataHandleTest extends TestCase
{
    public function test_rule_evaluate_accepts_data_handle(): void
    {
        $engine = new Engine();
        $rule = $engine->compile('{"*":[{"var":"x"},2]}');
        $data = new DataHandle('{"x":21}');
        self::assertSame('42', $rule->evaluate($data));
        // Not consumed by evaluation — reusable.
        self::assertSame('42', $rule->evaluate($data));
        // The string overload keeps working side by side.
        self::assertSame('42', $rule->evaluate('{"x":21}'));
    }

    public function test_session_evaluate_accepts_data_handle(): void
    {
        $engine = new Engine();
        $rule = $engine->compile('{"+":[{"var":"a"},{"var":"b"}]}');
        $session = $engine->openSession();
        $data = new DataHandle('{"a":40,"b":2}');
        self::assertSame('42', $session->evaluate($rule, $data));
        self::assertSame('42', $session->evaluate($rule, '{"a":40,"b":2}'));
    }

    public function test_data_handle_is_engine_independent(): void
    {
        $data = new DataHandle('{"x":7}');
        $a = new Engine();
        $b = new Engine(templating: true);
        self::assertSame('8', $a->compile('{"+":[{"var":"x"},1]}')->evaluate($data));
        self::assertSame('6', $b->compile('{"-":[{"var":"x"},1]}')->evaluate($data));
    }

    public function test_allocated_bytes_is_positive_and_zero_after_close(): void
    {
        $data = new DataHandle('{"x":[1,2,3,4,5]}');
        self::assertGreaterThan(0, $data->allocatedBytes());
        $data->close();
        self::assertSame(0, $data->allocatedBytes());
        $data->close(); // idempotent
    }

    public function test_using_closed_handle_throws(): void
    {
        $engine = new Engine();
        $rule = $engine->compile('{"var":"x"}');
        $data = new DataHandle('{"x":1}');
        $data->close();
        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessage('DataHandle has been closed');
        $rule->evaluate($data);
    }

    public function test_malformed_json_throws_ParseException(): void
    {
        try {
            new DataHandle('{"x":');
            self::fail('expected ParseException');
        } catch (ParseException $ex) {
            self::assertSame('ParseError', $ex->errorType);
            self::assertNotSame('', $ex->getMessage());
        }
    }
}

File: bindings/jvm/src/test/java/com/goplasmatic/datalogic/DataHandleTest.java
/* SPDX-License-Identifier: Apache-2.0 */
package com.goplasmatic.datalogic;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import static org.junit.jupiter.api.Assertions.*;

class DataHandleTest {

    @Test
    void parse_then_evaluate_via_rule_and_session() {
        try (Engine engine = new Engine();
             Rule rule = engine.compile("{\"*\":[{\"var\":\"x\"},2]}");
             Session session = engine.openSession();
             DataHandle data = DataHandle.parse("{\"x\":21}")) {
            // Not consumed by evaluation: use it repeatedly, on both paths.
            assertEquals("42", rule.evaluate(data));
            assertEquals("42", rule.evaluate(data));
            assertEquals("42", session.evaluate(rule, data));
            assertEquals("42", session.evaluate(rule, data));
        }
    }

    @Test
    void handle_is_engine_independent() {
        try (Engine a = new Engine();
             Engine b = new Engine();
             Rule ruleA = a.compile("{\"var\":\"x\"}");
             Rule ruleB = b.compile("{\"+\":[{\"var\":\"x\"},1]}");
             DataHandle data = DataHandle.parse("{\"x\":41}")) {
            assertEquals("41", ruleA.evaluate(data));
            assertEquals("42", ruleB.evaluate(data));
        }
    }

    @Test
    void malformed_json_throws_ParseException() {
        ParseException ex = assertThrows(ParseException.class, () -> DataHandle.parse("not-json{{"));
        assertEquals("ParseError", ex.errorType());
        assertNotNull(ex.getMessage());
    }

    @Test
    void allocated_bytes_is_positive() {
        try (DataHandle data = DataHandle.parse("{\"x\": [1, 2, 3, \"four\"]}")) {
            assertTrue(data.allocatedBytes() > 0);
        }
    }

    @Test
    void closed_handle_throws_IllegalStateException() {
        DataHandle data = DataHandle.parse("{\"x\":1}");
        data.close();
        try (Engine engine = new Engine();
             Rule rule = engine.compile("{\"var\":\"x\"}")) {
            assertThrows(IllegalStateException.class, () -> rule.evaluate(data));
        }
        data.close(); // double-close is a no-op
    }

    @Test
    void one_handle_shared_across_threads() throws Exception {
        try (Engine engine = new Engine();
             Rule rule = engine.compile("{\"*\":[{\"var\":\"x\"},2]}");
             DataHandle data = DataHandle.parse("{\"x\":21}")) {
            int threads = 4;
            ExecutorService pool = Executors.newFixedThreadPool(threads);
            try {
                List<Future<Boolean>> results = new ArrayList<>();
                for (int t = 0; t < threads; t++) {
                    results.add(pool.submit(() -> {
                        // One session per thread; rule + data are shared.
                        try (Session session = engine.openSession()) {
                            for (int i = 0; i < 500; i++) {
                                if (!"42".equals(session.evaluate(rule, data))) return false;
                                if (session.evaluateLong(rule, data) != 42L) return false;
                            }
                        }
                        return true;
                    }));
                }
                for (Future<Boolean> f : results) {
                    assertTrue(f.get(), "worker saw a wrong evaluation result");
                }
            } finally {
                pool.shutdown();
            }
        }
    }
}
